compiler/src/lnntoamm/Fn.ts

```diff
@@ -219,22 +247,43 @@ export default class Fn {
       }
       stmt.inline(amm);
     }
+    this.retTy.resetTemp();
     this.params.forEach((param) => param.unassign());
   }
 
-  resultTyFor(argTys: Type[], scope: Scope): Type | null {
-    let res: Type | null = null;
+  resultTyFor(
+    argTys: Type[],
+    expectResTy: Type,
+    scope: Scope,
+  ): [Type[], Type] | null {
+    let res: [Type[], Type] | null = null;
     try {
-      this.params.forEach((param, ii) =>
-        param.ty.tempConstrain(argTys[ii], scope),
-      );
-      res = this.retTy.instance();
-    } catch (_e) {
+      this.params.forEach((param, ii) => {
+        return param.ty.tempConstrain(argTys[ii], scope);
+      });
```

Comment:
Nit: Single-line functions like these have tended to go without the curly braces, so I am adding a suggested change to make it more consistent with the rest of the code in here.

```suggestion
      this.params.forEach((param, ii) => param.ty.tempConstrain(argTys[ii], scope));
```

Reply:
yeah, i had to do a block at one point for debugging purposes, but i'll change it back